homeassistant/components/hafas/sensor.py

```diff
@@ -0,0 +1,169 @@
+"""Sensor for HaFAS."""
+
+from __future__ import annotations
+
+from datetime import timedelta
+from enum import Enum
+from typing import Any
+
+from pyhafas import HafasClient
+from pyhafas.profile import DBProfile, VSNProfile
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA, SensorEntity
+from homeassistant.const import CONF_OFFSET
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
+import homeassistant.util.dt as dt_util
+
+DOMAIN = "hafas"
+
+CONF_PROFILE = "profile"
+CONF_DESTINATION = "to"
+CONF_START = "from"
+DEFAULT_OFFSET = timedelta(minutes=0)
+CONF_ONLY_DIRECT = "only_direct"
+DEFAULT_ONLY_DIRECT = False
+
+ICON = "mdi:train"
+
+SCAN_INTERVAL = timedelta(minutes=2)
+
+
+class Profile(Enum):
+    """Enum of HaFAS profile type."""
+
+    DB = 0
+    VSN = 1
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_PROFILE): cv.enum(Profile),
+        vol.Required(CONF_DESTINATION): cv.string,
+        vol.Required(CONF_START): cv.string,
+        vol.Optional(CONF_OFFSET, default=DEFAULT_OFFSET): cv.time_period,
+        vol.Optional(CONF_ONLY_DIRECT, default=DEFAULT_ONLY_DIRECT): cv.boolean,
+    }
+)
+
+
+def setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    add_entities: AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Set up the HaFAS."""
+    profile = config.get(CONF_PROFILE)
+    start = config.get(CONF_START)
+    destination = config.get(CONF_DESTINATION)
+    offset = config.get(CONF_OFFSET)
+    only_direct = config.get(CONF_ONLY_DIRECT)
+
+    add_entities([HaFAS(profile, start, destination, offset, only_direct)], True)
+
+
+class HaFAS(SensorEntity):
+    """Implementation of a HaFAS sensor."""
+
+    def __init__(self, profile, start, goal, offset, only_direct):
+        """Initialize the sensor."""
+        self._name = f"{start} to {goal}"
+        self.offset = offset
+        self.only_direct = only_direct
+
+        self.client = None
+        if profile == Profile.DB:
+            self.client = HafasClient(DBProfile())
+        elif profile == Profile.VSN:
+            self.client = HafasClient(VSNProfile())
+
+        origins = self.client.locations(start)
+        self.origin = origins[0]
+
+        destinations = self.client.locations(goal)
+        self.destination = destinations[0]
+
+        self.update()
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self) -> str:
+        """Return the icon for the frontend."""
+        return ICON
+
+    @property
+    def native_value(self) -> str:
```

Comment:
I think ideally, some of this logic should be moved to the underlying python library (pyhafas)

Reply:
Not really sure, which logic you are referencing here

Reply:
The string generation and such in the `native_value` function. That could be done in the python library since it's not inherently specific to home assistant

Reply:
You are right, that this is not inherently specific to home assistant, but it wouldn't fit into `pyhafas`. The library seems to focus on being a source of data and not a fronted. Therefore, it does not output any data, other than received from the endpoints.